### PR TITLE
Daml-LF. dynamic serializability proof.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR is a proof of concept (which does not aim to be merged) that shows how could be added dynamic proof that a term is serialize. 

_For the sake of simplicity, this proof is limited to `Int64`, `List`, and serializable user-defined type._ 

Conceptually, we have a new type `Serializable τ` for typing serializability proof. 
The correctness/soundness is obtain by providing a limited set of builtin functions that are the only one able to create such proof. 

Then we define a generic equality function (for serializable type) that take in parameter such a proof to ensure the type is serializable.

From a typeclass perspective the proof can be thought as a builtin dictionary that provide the equal function. 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/archive/da/daml_lf_1.proto [hunk 1]
```diff
@@ -825,6 +832,8 @@ message Expr {
     // Extract the given type from Any or return None on type-mismatch ('ExpFromAny')
     // *Available in versions >= 1.dev*
     FromAny from_any = 30;
+
+    TypeConName data_is_serializable = 999;
```
Comment (hurryabit) [inline daml-lf/archive/da/daml_lf_1.proto]: What's the semantics of this?

File: compiler/damlc/daml-stdlib-src/DA/Serializable.daml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+-- | LF.TextMap - A map is an associative array data type composed of a
+-- collection of key/value pairs such that, each possible key appears
+-- at most once in the collection.
+module DA.Serializable (Serializable, (===), isSerializable, force1, force2) where
+
+import DA.Internal.LF qualified as LF
+
+class Serializable a where
+  isSerializable : LF.Serializable a
+  (===): a -> a -> Bool
```
Comment (hurryabit) [inline compiler/damlc/daml-stdlib-src/DA/Serializable.daml]: This should be moved to the top-level:
```haskell
isSerializable : Serializable a => a -> a -> Bool
isSerializable = LF.equal isSerializble
```

File: compiler/damlc/daml-stdlib-src/DA/Serializable.daml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+instance Serializable Int where
+  isSerializable  = LF.intIsSerializable
+
+instance (Serializable a) => Serializable [a] where
+  isSerializable = LF.listIsSerializable isSerializable
+
+
+force1: (LF.Serializable a -> LF.Serializable b) -> (LF.Serializable a -> LF.Serializable b)
+force1 x = x
```
Comment (hurryabit) [inline compiler/damlc/daml-stdlib-src/DA/Serializable.daml]: What's the purpose of this?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 4]
```diff
@@ -282,6 +283,10 @@ data BuiltinExpr
   | BETrace                      -- :: forall a. Text -> a -> a
   | BEEqualContractId            -- :: forall a. ContractId a -> ContractId a -> Bool
   | BECoerceContractId           -- :: forall a b. ContractId a -> ContractId b
+
+  | BEEqualSerializable          -- :: ∀a. Serializable a -> a -> a -> Bool,
+  | BEIntIsSerializable            -- :: Serializable Int
+  | BEListIsSerializable           -- :: ∀a. Serializable a -> Serializable [a],
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: How is this going to work for user-defined types, e.g., let’s say I define my own type isomorphic to the builtin list?

File: daml-lf/archive/da/daml_lf_1.proto (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]:
```
data Mod:T α₁ ⋯ αₙ = ... is a serializable user-define type
------------------------------------------------------------------------------------------------------
data_is_serializable @Mod:T : ∀ α₁ ⋯ αₙ. Serializable α₁ → ⋯ → Serializable αₙ → Serializable (Mod:T α₁ ⋯ αₙ)
``

Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]: This is to built proof builtin function for user-defined type.

Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]:
```
data Mod:T α₁ ⋯ αₙ = ... is a serializable user-define type
------------------------------------------------------------------------------------------------------
data_is_serializable : ∀ α₁ ⋯ αₙ. Serializable α₁ → ⋯ → Serializable αₙ → Serializable (Mod:T α₁ ⋯ αₙ)
```

Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]: It is use to build proof building function for user-defined types

File: compiler/damlc/daml-stdlib-src/DA/Serializable.daml (hunk 3 above)
Comment (remyhaemmerle-da) [reply compiler/damlc/daml-stdlib-src/DA/Serializable.daml]: It is a hack to force a polymorphic type to have the shape 
`(LF.Serializable a -> LF.Serializable b)`

File: daml-lf/archive/da/daml_lf_1.proto (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]: I use that to force the type of `(primitive @"BEDataIsSerializable")` to have a type of the good shape when converting to daml-lf (in primitive.hs)

File: compiler/damlc/daml-stdlib-src/DA/Serializable.daml (hunk 3 above)
Comment (remyhaemmerle-da) [reply compiler/damlc/daml-stdlib-src/DA/Serializable.daml]: I use it to force the type of `(primitive @"BEDataIsSerializable)"` to have the proper type so I can convert porperly in daml-lf in primitive.hs

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs (hunk 4 above)
Comment (remyhaemmerle-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: See my first answer to @hurryabit

File: daml-lf/archive/da/daml_lf_1.proto (hunk 1 above)
Comment (cocreature) [reply daml-lf/archive/da/daml_lf_1.proto]: So I can’t get equality if I have a phantom type parameter that is not serializable?

**Pull request closed without merging**